pymchelper/averaging.py

```diff
@@ -0,0 +1,214 @@
+
+    def update(self, value: Union[float, ArrayLike], weight: float = 1.0, **kwargs):
+        """
+        Update the state of the aggregator with new data.
+        Note that the weights are so called "reliability weights", not frequency weights.
+        If unsure put here the number of histories from the file if you are aggregating data from multiple files.
+        """
+        if weight < 0:
+            raise ValueError("Weight must be non-negative")
+
+        # first pass initialization
+        if not self.updated:
+            self.data = value * 0
+            self._accumulator_S = value * 0
+
+        # W_n = W_{n-1} + w_n
+        self.total_weight += weight
+        self._total_weight_squared += weight**2
+
+        mean_old = self.data
+        # mu_n = (1 - w_n / W_n) * mu_{n-1} + (w_n / W_n) * x_n
+        # or in other words:
+        # mu_n - mu_{n-1} = (w_n / W_n) * (x_n - mu_{n-1})
+        self.data += (weight / self.total_weight) * (value - mean_old)
+
+        self._accumulator_S += weight * (value - self.data) * (value - mean_old)
+
+        self._updated = True
+        logging.debug("Updated aggregator with value %s and weight %s", value, weight)
+
+    @property
+    def mean(self):
+        """Weighted mean of the sample"""
+        return self.data
+
+    @property
+    def variance_population(self):
+        """Biased estimate of the variance"""
+        return self._accumulator_S / self.total_weight
+
+    @property
+    def variance_sample(self):
+        """
+        Unbiased estimate of the variance.
+        The bias of the weighted estimator if (1 - sum w_i^2 / W_n^2), or in other words:
+        1 - "sum of squares of weights" / "square of sum of weights".
+        For all equal weights the bias is 1 - n * w^2/((n * w)^2) = 1 - 1/n which
+        leads to the well known formula for the sample variance.
+        Here we use the weighted version of the formula.
+        """
+        return self._accumulator_S / (self.total_weight - (self._total_weight_squared / self.total_weight))
+
+    @property
+    def stddev(self):
+        """Standard deviation of the sample"""
+        return np.sqrt(self.variance_sample)
+
+    @property
+    def stderr(self):
+        """
+        Standard error of the sample.
+        For weighted data it is calculated as:
+        stddev * sqrt(sum w_i^2) / sum w_i
+        For equal weights it reduces via sqrt( n * w^2) / (n * w) = stddev / sqrt(n)
+        """
+        return self.stddev * np.sqrt(self._total_weight_squared) / self.total_weight
+
+    def error(self, **kwargs):
+        """Error calculation function, can be used to calculate standard deviation or standard error."""
+        logging.debug("Calculating error with kwargs: %s", kwargs)
+        if 'error_type' in kwargs:
+            if kwargs['error_type'] == 'stddev':
+                return self.stddev
+            if kwargs['error_type'] == 'stderr':
+                return self.stderr
+        return None
```

Comment:
is error_type always set to either stddev or stderr? (maybe throw exception with clear error message)

Reply:
We could have user which requested no error to be calculated. The error type is defined here:
https://github.com/DataMedSci/pymchelper/blob/93cfda176bce289504c2a72bb1122166f6e080ac/pymchelper/estimator.py#L12-L19

if no error is requested then I think we could safely return None. 
Anyway good point - I will improve the code

Reply:
Fixed in f82d04e4d5f9772b12485ad29571f06ac2196b9b

Reply:
second last return None is redundant